woocommerce/woocommerce-admin
# Remote inbox notification delivery

Fixes #3878

p90Yrv-1Hr-p2

The shop polls data sources (TBC) for a JSON file containing specs. Each spec contains a number of rules and information about an admin note that would be created if the rules pass the engine.

Up for conversation:

- how often to poll the JSON data source(s)
- how often to run the engine
- ideas for other rules
- everything else ;-)

There's also a P2 post to discuss what would be needed for a [server-side implementation](https://teamisotope.wordpress.com/2020/05/05/rule-based-notifications-server-side-implementation/).

### JSON description

A test JSON specs file is available from [here](https://gist.githubusercontent.com/becdetat/f17e9617e5ca1211d0579cf079862905/raw/7f0e44287b21cbf53a632887a805a414137000ce/rinds-specs.json).

#### JSON file:

```
[
    <spec>,
    ...
]
```

#### Spec:

```
{
    "slug": "ad-for-automate-woo-2020-04-20",
    "type": "info",
    "icon": "info",
    "status": "unactioned",
    "is_snoozable": 0,
    "source": "woocommerce.com",
    "locales": [
        <Locale>,
        ...
    ],
    "actions": [
        <Action>,
        ...
    ],
    "rules": [
        <Rule>,
        ...
    ]
```

#### Locale

```
{
    "locale": "en_US",
    "title": "Ad for Automate Woo",
    "content": "Hi There! This is an ad for Automate Woo."
},
{
    "locale": "en_AU",
    "title": "Ad for Automate Woo, in en-AU.",
    "content": "en-AU content"
},
{
    "locale": "fr_FR",
    "title": "Annonce pour automatiser woo",
    "content": "Salut! Ceci est une publicité pour Automate Woo."
}
```

#### Action

```
{
    "name": "install-automate-woo",
     "locales": [
        <ActionLocale>,
        ...
    ],
    "url": "?page=automatewoo-dashboard",
    "url_is_admin_query": true,
    "is_primary": true,
    "status": "actioned"
}
```

#### ActionLocale

```
{
    "locale": "en_US",
    "label": "Install"
}
```

#### Rule

Rules in an array are executed as an AND operation. If there are no rules in the array the result is `false`.

Plugins activated:

```
{
    "type": "plugins_activated",
    "plugins": [
        "plugin-slug-1',
        "plugin-slug-2"
    ]
}
```

Publish after time:

```
{
    "type": "publish_after_time",
    "publish_after": "2020-04-22 00:00:00"
}
```

Publish before time:

Note that using both `publish_after_time` and `publish_before_time` allows timeboxing a note which could be useful for promoting a sale.

```
{
    "type": "publish_before_time",
    "publish_before": "2020-04-22 00:00:00"
}
```

Not:

Note that the rules in `operand` get ANDed together into a single boolean.

```
{
    "type": "not",
    "operand": [
        <Rule>,
        ...
    ]
}
```

Or:

Note that the rules in `left_operand` and `right_operand` get ANDed together into a single boolean value for the operand.

```
{
    "type": "or",
    "operands": [
        [
            <Rule>,
            ...
        ],
        [
            <Rule>,
            ...
        ]
    ]
}
```

alternatively:

```
{
    "type": "or",
    "operands": [
        <Rule>,
        ...
    ]
}
```

Fail:

This just returns a `false` value.

```
{
    "type": "fail"
}
```
Plugin version:

This compares the installed version of the plugin to the required version, using the operator. If the plugin isn't activated this returns false.


```
{
    "type": "plugin_version",
    "plugin": "jetpack",
    "version": "8.4.1",
    "operator": "="
}
```

### Detailed test instructions:

- Switch to this PR's branch.
- Note that this adds a feature flag so `npm start` will have to be re-run
- Install the WP Crontrol plugin to access cron jobs via Tools -> Cron events. Running the `wc_admin_daily` hook triggers a poll, and also triggers a run of the engine at the moment. This will probably be separated out to different events though as the engine should probably be run more frequently than a daily poll. Activating and deactivating plugins also triggers an engine run (but not a poll).
- To check which specs have triggered, use a page that shows the admin notes (such as the WooCommerce dashboard).
- Copy [this gist](https://gist.githubusercontent.com/becdetat/f17e9617e5ca1211d0579cf079862905/raw/99495c525c54663f6a0b90edff81e9ccd5b72afb/rinds-specs.json) somewhere it will be picked up by a local web server. For example, if you're using vvv, `<vvv root>/www/wordpress-two/public_html/rinds-specs.json` will be available at http://two.wordpress.test/rinds-specs.json. (@mattsherman updated link to gist on 2020-05-27)
- Update `DataSourcePoller::DATA_SOURCES` to poll the above file
- In the spec, set the `ad-for-automate-woo-2020-04-22` spec's rule's `publish_after` value to some time in the future.
- If you have WooCommerce Services or Jetpack activated, deactivate them
- Delete all records in the `wp-wc-admin-notes` table
- Trigger a poll

Let's get testing:

- set the `ad-for-automate-woo-2020-04-22` spec's rules's `publish_after_time` time to some time in the past; make sure the `publish_before_time` time is set to some time in the future (@mattsherman updated this on 2020-05-27)
- trigger a poll
- check that the new note appears 

- install and activate both Jetpack and WooCommerce Services
- check that the `more-action-for-jp-wcs` note appears

- deactivate Jetpack
- check that the `NOT-test` note appears and the `ad-for-automate-woo-2020-04-22` note disappears

- test that the `OR-test` note appears if either (or both) of the Jetpack or WooCommerce Services plugins are activated

### Changelog Note:

Enhancement: Remote inbox notification delivery
